R/distribution_formulas.R

```diff
@@ -0,0 +1,192 @@
+#'   option listed, i.e., `"index"` is the default.
+#' @param module An identifier to a C++ fleet module that is linked to the data
+#'   of interest.
+#' @param family A description of the error distribution and link function to
+#'   be used in the model.
+#' @param sd A list of length two. The first entry, `"value"`, stores the
+#'   initial values for the relevant standard deviations. The second entry,
+#'   `"estimated"` is a vector of booleans indicating whether or not
+#'   standard deviation is estimated.
+#' @export
+new_data_distribution <- function(
+  module,
+  family,
+  sd = list(value = 1, estimated = FALSE),
+  data_type = c("index", "cpue", "agecomp", "lengthcomp")
+) {
+  data_type <- match.arg(data_type)
+  families <- c("lognormal", "gaussian", "multinomial")
+  if (family[["family"]] == "normal") {
+    stop("use family = gaussian() instead")
+  }
+  if (!(family[["family"]] %in% families)) {
+    stop("FIMS currently does not offer this distribution.")
+  }
+  if ((data_type == "agecomp" || data_type == "lengthcomp") &&
+      (family[["family"]] == "lognormal" || family[["family"]] == "gaussian")) {
+    stop("Did you mean family = multinomial()?")
+  }
+  if ((data_type == "index" || data_type == "cpue") &&
+      family[["family"]] == "multinomial") {
+    stop("Multinomial is not available for index or CPUE data")
+  }
+
+  if (data_type == "index" || data_type == "cpue") {
+    obs_id_name <- "observed_index_data_id"
+  }
+  if (data_type == "agecomp" || data_type == "lengthcomp") {
+    obs_id_name <- "observed_agecomp_data_id"
+  }
+  if (family[["family"]] == "lognormal") {
+    new_module <- new(TMBDlnormDistribution)
+    new_module$log_logsd <- new(
+      ParameterVector,
+      log(sd$value),
+      length(sd$value)
+    )
+    new_module$log_logsd$set_all_estimable(sd$estimated)
+    if (family[["link"]] == "log") {
+      expected <- "log_expected_index"
+    }
+    if (family[["link"]] == "identity") {
+      expected <- "expected_index"
+    }
+  }
+
+  if (family[["family"]] == "gaussian") {
+    new_module <- new(TMBDnormDistribution)
+    new_module$log_sd <- new(ParameterVector, log(sd$value), length(sd$value))
+    new_module$log_sd$set_all_estimable(sd$estimated)
+    if (family[["link"]] == "log") {
+      expected <- "log_expected_index"
+    }
+    if (family[["link"]] == "identity") {
+      expected <- "expected_index"
+    }
+  }
+
+  if (family[["family"]] == "multinomial") {
+    new_module <- new(TMBDmultinomDistribution)
+    expected <- "proportion_catch_numbers_at_age"
+  }
+  if (data_type == "index" || data_type == "cpue") {
+    new_module$set_observed_data(module$GetObservedIndexDataID())
+  }
+  if (data_type == "agecomp") {
+    new_module$set_observed_data(module$GetObservedAgeCompDataID())
+  }
+
+  new_module$set_distribution_links("data", module$field(expected)$get_id())
+  return(new_module)
+}
+
+#' Sets up a new distribution for a process
```

Comment:
Maybe add documentation for the return value, similar to that of `new_data_distribution()`?
Maybe provide more details about valid options for `par` and how they relate to the `module`?
It would be great to add a `\dontrun{}` example.

Reply:
added requested documentation